src/main/kotlin/io/kotlintest/Spec.kt

```diff
@@ -71,99 +39,27 @@ abstract class Spec {
-   * executed.
+   * To continue execution of this spec class, you must invoke the chain function. If you don't
+   * want to continue with the execution of the spec, then do not invoke the chain function.
    */
-  protected open fun interceptSpec(context: Spec, spec: () -> Unit) {
-    spec()
-  }
-
-  private fun runOneInstancePerTest(notifier: RunNotifier) {
-    val testCount = rootTestSuite.testCasesIncludingChildren().size
-    for (testCaseIndex in (0 until testCount)) {
-      val instance = javaClass.newInstance()
-      val testCase = instance.rootTestSuite[testCaseIndex]
-      runTest(instance, testCase, notifier)
-    }
-  }
-
-  private fun runSharedInstance(notifier: RunNotifier) {
-    val testCases = rootTestSuite.testCasesIncludingChildren()
-    testCases.forEach { runTest(this, it, notifier) }
-  }
-
-  // TODO beautify
-  private fun runTest(
-      spec: Spec,
-      testCase: TestCase,
-      notifier: RunNotifier) {
-    if (testCase.isActive) {
-      val executor =
-          if (testCase.config.threads < 2) Executors.newSingleThreadExecutor()
-          else Executors.newFixedThreadPool(testCase.config.threads)
-      notifier.fireTestStarted(testCase.description)
-      val initialInterceptor = { context: TestCaseContext, testCase: () -> Unit ->
-        interceptTestCase(context, { testCase() })
-      }
-      val interceptorChain = createInterceptorChain(testCase.config.interceptors, initialInterceptor)
-      val testCaseContext = TestCaseContext(spec, testCase)
-      val results = ArrayList<Future<Any>>()
-      for (j in 1..testCase.config.invocations) {
-        val callable = Callable {
-          try {
-            interceptorChain(testCaseContext, { testCase.test() })
-          } catch (exception: Throwable) {
-            Failure(testCase.description, exception)
-          }
-        }
-        results.add(executor.submit(callable))
-      }
-      executor.shutdown()
-      val timeout = testCase.config.timeout
-      val terminated = executor.awaitTermination(timeout.amount, timeout.timeUnit)
-      results.forEach {
-        val result = it.get()
-        if (result is Failure) {
-          notifier.fireTestFailure(result)
-        }
-      }
-      if (!terminated) {
-        val failure = Failure(description(), TestTimedOutException(timeout.amount, timeout.timeUnit))
-        notifier.fireTestFailure(failure)
-      }
-      notifier.fireTestFinished(testCase.description)
-    } else {
-      notifier.fireTestIgnored(testCase.description)
-    }
-  }
-
-  private fun <CONTEXT> createInterceptorChain(
-      interceptors: Iterable<(CONTEXT, () -> Unit) -> Unit>,
-      initialInterceptor: (CONTEXT, () -> Unit) -> Unit): (CONTEXT, () -> Unit) -> Unit {
-    return interceptors.reversed().fold(initialInterceptor) { a, b ->
-      {
-        context: CONTEXT, testCase: () -> Unit -> b(context, { a.invoke(context, { testCase() }) })
-      }
-    }
+  internal open fun interceptSpec(chain: () -> Unit) {
```

Comment:
I think `chain` is misleading here, since the intercepted thing is primarily the spec itself. The fact that the spec itself could be decorated with other interceptors is not as important here. From the callers point of view, it is the spec. --> rename to `spec`.